File: leetcode/152.maximum-product-subarray.py
# @leet start
import math

class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        # Go through with a forward iteration. Conceptually, nums is
        # broken up into chunks with separators 0. Within each chunk,
        # we'll keep track of the current maximum and minimum of
        # contiguous numbers we can get a product of.

        # NOTE: I got very close to this on my own, but ultimately
        # needed used Neetcode to figure out exactly how to do this
        current_max = 1
        current_min = 1
        best = -math.inf

        for num in nums:
            # If num == 0 we move onto the next chunk
            if num == 0:
                current_max = 1
                current_min = 1

                best = max(best, 0)

                continue

            # We'll set current max to this after we're done using it
            # for current_min
            new_current_max = max(current_max * num, current_min * num, num)

            current_min = min(current_max * num, current_min * num, num)
            current_max = new_current_max

            best = max(best, current_max)

        return best
# @leet end

File: leetcode/152.maximum-product-subarray.js
// @leet start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
  let curr = 1;
  let startNeg = null;
  let bestNeg = null;
  let bestPos = null;
  let result = nums[0];

  const updateResult = (atZero) => {
    if (atZero) {
      result = Math.max(result, 0);
    }

    if (bestNeg !== null) {
      result = Math.max(result, bestNeg / startNeg);
    }

    if (bestPos !== null) {
      result = Math.max(result, bestPos);
    }
  };

  for (const num of nums) {
    if (num === 0) {
      updateResult(true);
      curr = 1;
      startNeg = null;
      bestNeg = null;
      bestPos = null;
    } else {
      curr *= num;

      if (curr < 0) {
        if (startNeg === null) {
          startNeg = curr;
        } else {
          bestNeg = Math.min(bestNeg, curr);
        }
      } else {
        // curr > 0
        bestPos = Math.max(bestPos, curr);
      }
    }
  }

  updateResult(false);

  return result;
};
// @leet end
